File: packages/python/src/ngtaxkit/pdf/invoice_pdf.py
"""Invoice PDF generation using fpdf2."""

from __future__ import annotations

from typing import Any

from . import _require_fpdf2
from .helpers import (
    COLOR_HEADER_BG, COLOR_LINE, COLOR_PRIMARY, COLOR_SECONDARY, COLOR_WHITE,
    CONTENT_WIDTH, PAGE_MARGIN, format_currency, format_date,
)


def _cur(amount: float) -> str:
    """Format currency for PDF output (latin-1 safe)."""
    return format_currency(amount, pdf_safe=True)


def to_pdf(inv: dict[str, Any]) -> bytes:
    """Generate an invoice PDF as bytes."""
    fpdf = _require_fpdf2()
    pdf = fpdf.FPDF(orientation="P", unit="mm", format="A4")
    pdf.set_auto_page_break(auto=True, margin=20)
    pdf.add_page()

    # -- Letterhead --
    pdf.set_font("Helvetica", "B", 16)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 8, inv["seller"]["name"], new_x="LMARGIN", new_y="NEXT")

    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, inv["seller"]["address"], new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"TIN: {inv['seller']['tin']}", new_x="LMARGIN", new_y="NEXT")
    if inv["seller"].get("vrn"):
        pdf.cell(0, 4, f"VRN: {inv['seller']['vrn']}", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(6)

    # -- TAX INVOICE banner --
    type_label = {"credit-note": "CREDIT NOTE", "debit-note": "DEBIT NOTE"}.get(inv["type"], "TAX INVOICE")
    pdf.set_fill_color(*COLOR_PRIMARY)
    pdf.set_text_color(*COLOR_WHITE)
    pdf.set_font("Helvetica", "B", 14)
    pdf.cell(CONTENT_WIDTH, 10, type_label, fill=True, new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- Invoice metadata --
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, f"Invoice #: {inv['invoice_number']}", new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"Issue Date: {format_date(inv['issue_date'])}", new_x="LMARGIN", new_y="NEXT")
    if inv.get("due_date"):
        pdf.cell(0, 4, f"Due Date: {format_date(inv['due_date'])}", new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"Currency: {inv['currency']}", new_x="LMARGIN", new_y="NEXT")
    if inv.get("purchase_order_ref"):
        pdf.cell(0, 4, f"PO Ref: {inv['purchase_order_ref']}", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- Bill To --
    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.cell(0, 5, "Bill To:", new_x="LMARGIN", new_y="NEXT")
    pdf.set_font("Helvetica", "", 9)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.cell(0, 4, inv["buyer"]["name"], new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, inv["buyer"]["address"], new_x="LMARGIN", new_y="NEXT")
    pdf.cell(0, 4, f"TIN: {inv['buyer']['tin']}", new_x="LMARGIN", new_y="NEXT")
    if inv["buyer"].get("vrn"):
        pdf.cell(0, 4, f"VRN: {inv['buyer']['vrn']}", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(6)

    # -- Line items table --
    col_widths = [60, 15, 30, 22, 25, 28]
    headers = ["Description", "Qty", "Unit Price", "VAT %", "VAT Amt", "Total"]

    pdf.set_fill_color(*COLOR_HEADER_BG)
    pdf.set_font("Helvetica", "B", 8)
    pdf.set_text_color(*COLOR_PRIMARY)
    for i, h in enumerate(headers):
        align = "L" if i == 0 else "R"
        pdf.cell(col_widths[i], 6, h, border=0, fill=True, align=align)
    pdf.ln()

    pdf.set_font("Helvetica", "", 8)
    pdf.set_text_color(*COLOR_SECONDARY)
    for item in inv["items"]:
        pdf.cell(col_widths[0], 5, item["description"][:40], align="L")
        pdf.cell(col_widths[1], 5, str(item["quantity"]), align="R")
        pdf.cell(col_widths[2], 5, _cur(item["unit_price"]), align="R")
        pdf.cell(col_widths[3], 5, f"{item['vat_rate'] * 100:.1f}%", align="R")
        pdf.cell(col_widths[4], 5, _cur(item["vat_amount"]), align="R")
        pdf.cell(col_widths[5], 5, _cur(item["line_total"]), align="R")
        pdf.ln()
        pdf.set_draw_color(*COLOR_LINE)
        pdf.line(pdf.l_margin, pdf.get_y(), pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
        pdf.ln(1)

    pdf.ln(4)

    # -- Totals --
    totals_x = pdf.l_margin + CONTENT_WIDTH - 70
    pdf.set_font("Helvetica", "", 9)
    for label, value in [("Subtotal:", inv["subtotal"]), ("Total VAT:", inv["total_vat"])]:
        pdf.set_x(totals_x)
        pdf.cell(35, 5, label, align="L")
        pdf.cell(35, 5, _cur(value), align="R", new_x="LMARGIN", new_y="NEXT")

    pdf.set_font("Helvetica", "B", 10)
    pdf.set_text_color(*COLOR_PRIMARY)
    pdf.set_x(totals_x)
    pdf.cell(35, 6, "TOTAL:", align="L")
    pdf.cell(35, 6, _cur(inv["total"]), align="R", new_x="LMARGIN", new_y="NEXT")
    pdf.ln(4)

    # -- VAT breakdown --
    if inv["vat_breakdown"]:
        pdf.set_font("Helvetica", "B", 10)
        pdf.set_text_color(*COLOR_PRIMARY)
        pdf.cell(0, 5, "VAT Breakdown", new_x="LMARGIN", new_y="NEXT")
        pdf.set_font("Helvetica", "", 8)
        pdf.set_text_color(*COLOR_SECONDARY)
        for entry in inv["vat_breakdown"]:
            rt = entry["rate_type"]
            rate_label = f"{entry['rate'] * 100:.1f}% ({rt.replace('-', ' ').title()})"
            pdf.cell(0, 4,
                f"{rate_label}  --  Taxable: {_cur(entry['taxable_amount'])}  |  VAT: {_cur(entry['vat_amount'])}",
                new_x="LMARGIN", new_y="NEXT")
        pdf.ln(4)

    # -- Notes --
    if inv.get("notes"):
        pdf.set_font("Helvetica", "", 8)
        pdf.set_text_color(*COLOR_SECONDARY)
        pdf.cell(0, 4, f"Notes: {inv['notes']}", new_x="LMARGIN", new_y="NEXT")
        pdf.ln(4)

    # -- Footer --
    pdf.set_draw_color(*COLOR_LINE)
    pdf.line(pdf.l_margin, pdf.get_y(), pdf.l_margin + CONTENT_WIDTH, pdf.get_y())
    pdf.ln(3)
    pdf.set_font("Helvetica", "", 7)
    pdf.set_text_color(*COLOR_SECONDARY)
    pdf.multi_cell(CONTENT_WIDTH, 3,
        "This invoice is issued in compliance with the Nigeria Tax Act (NTA) 2025 and VAT Act as amended. "
        "VAT is charged at the applicable rate per NTA 2025 Schedule 1 (zero-rated), Schedule 2 (exempt), "
        "or the standard rate of 7.5%. All amounts are in Nigerian Naira (NGN) unless otherwise stated.")
    pdf.ln(2)
    pdf.cell(CONTENT_WIDTH, 3,
        f"Generated by ngtaxkit -- Invoice {inv['invoice_number']} | {inv['seller']['tin']}",
        align="C")

    return pdf.output()

File: packages/typescript/src/pdf/invoice-pdf.ts
// ─── Invoice PDF Generation (Layer 2) ────────────────────────────────────────
// Generates professional PDF invoices using pdfkit — no headless browser required.
// Returns a Buffer, performs no file I/O. Serverless-compatible.

import PDFDocument from 'pdfkit';
import type { Invoice } from '../invoice';

// ─── Formatting Helpers ──────────────────────────────────────────────────────

/** Format a number as Nigerian Naira: ₦X,XXX,XXX.XX */
function formatCurrency(amount: number): string {
  const fixed = Math.abs(amount).toFixed(2);
  const [whole, decimal] = fixed.split('.');
  const withCommas = whole.replace(/\B(?=(\d{3})+(?!\d))/g, ',');
  const sign = amount < 0 ? '-' : '';
  return `${sign}\u20A6${withCommas}.${decimal}`;
}

/** Format ISO date (YYYY-MM-DD) as DD/MM/YYYY */
function formatDate(iso: string): string {
  const parts = iso.split('-');
  if (parts.length !== 3) return iso;
  return `${parts[2]}/${parts[1]}/${parts[0]}`;
}

// ─── Layout Constants ────────────────────────────────────────────────────────

const PAGE_MARGIN = 50;
const PAGE_WIDTH = 595.28; // A4
const CONTENT_WIDTH = PAGE_WIDTH - PAGE_MARGIN * 2;

const FONT_BOLD = 'Helvetica-Bold';
const FONT_REGULAR = 'Helvetica';

const COLOR_PRIMARY = '#1a1a2e';
const COLOR_SECONDARY = '#555555';
const COLOR_HEADER_BG = '#f0f0f0';
const COLOR_LINE = '#cccccc';

// ─── PDF Generation ──────────────────────────────────────────────────────────

/**
 * Generate a PDF invoice as a Buffer.
 * Renders: letterhead, TAX INVOICE header, buyer details, line items table,
 * VAT breakdown, payment instructions, and footer with legal references.
 */
export function toPDF(invoice: Invoice): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: 'A4',
      margin: PAGE_MARGIN,
      info: {
        Title: `Invoice ${invoice.invoiceNumber}`,
        Author: invoice.seller.name,
      },
    });

    const chunks: Uint8Array[] = [];
    doc.on('data', (chunk: Uint8Array) => chunks.push(chunk));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', (err: Error) => reject(err));

    // ── Letterhead Area ──
    renderLetterhead(doc, invoice);

    // ── TAX INVOICE Header ──
    renderInvoiceHeader(doc, invoice);

    // ── Buyer Details ──
    renderBuyerDetails(doc, invoice);

    // ── Line Items Table ──
    renderLineItems(doc, invoice);

    // ── VAT Breakdown Summary ──
    renderVatBreakdown(doc, invoice);

    // ── Payment Instructions ──
    renderPaymentInstructions(doc, invoice);

    // ── Footer ──
    renderFooter(doc, invoice);

    doc.end();
  });
}


// ─── Render Sections ─────────────────────────────────────────────────────────

function renderLetterhead(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  doc
    .font(FONT_BOLD)
    .fontSize(16)
    .fillColor(COLOR_PRIMARY)
    .text(invoice.seller.name, PAGE_MARGIN, PAGE_MARGIN);

  doc
    .font(FONT_REGULAR)
    .fontSize(9)
    .fillColor(COLOR_SECONDARY)
    .text(invoice.seller.address, PAGE_MARGIN, doc.y + 2)
    .text(`TIN: ${invoice.seller.tin}`);

  if (invoice.seller.vrn) {
    doc.text(`VRN: ${invoice.seller.vrn}`);
  }

  doc.moveDown(1);
}

function renderInvoiceHeader(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  const y = doc.y;

  // "TAX INVOICE" banner
  doc
    .rect(PAGE_MARGIN, y, CONTENT_WIDTH, 28)
    .fill(COLOR_PRIMARY);

  doc
    .font(FONT_BOLD)
    .fontSize(14)
    .fillColor('#ffffff')
    .text('TAX INVOICE', PAGE_MARGIN + 10, y + 7, { width: CONTENT_WIDTH - 20 });

  doc.y = y + 38;

  // Invoice metadata
  const metaY = doc.y;
  doc
    .font(FONT_REGULAR)
    .fontSize(9)
    .fillColor(COLOR_SECONDARY);

  const typeLabel = invoice.type === 'credit-note' ? 'CREDIT NOTE' :
                    invoice.type === 'debit-note' ? 'DEBIT NOTE' : 'INVOICE';

  doc.text(`${typeLabel} #: ${invoice.invoiceNumber}`, PAGE_MARGIN, metaY);
  doc.text(`Issue Date: ${formatDate(invoice.issueDate)}`, PAGE_MARGIN, doc.y + 1);
  if (invoice.dueDate) {
    doc.text(`Due Date: ${formatDate(invoice.dueDate)}`, PAGE_MARGIN, doc.y + 1);
  }
  doc.text(`Currency: ${invoice.currency}`, PAGE_MARGIN, doc.y + 1);
  if (invoice.purchaseOrderRef) {
    doc.text(`PO Ref: ${invoice.purchaseOrderRef}`, PAGE_MARGIN, doc.y + 1);
  }

  doc.moveDown(1);
}

function renderBuyerDetails(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  doc
    .font(FONT_BOLD)
    .fontSize(10)
    .fillColor(COLOR_PRIMARY)
    .text('Bill To:', PAGE_MARGIN, doc.y);

  doc
    .font(FONT_REGULAR)
    .fontSize(9)
    .fillColor(COLOR_SECONDARY)
    .text(invoice.buyer.name)
    .text(invoice.buyer.address)
    .text(`TIN: ${invoice.buyer.tin}`);

  if (invoice.buyer.vrn) {
    doc.text(`VRN: ${invoice.buyer.vrn}`);
  }

  doc.moveDown(1);
}

function renderLineItems(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  // Column definitions: [label, x-offset, width, align]
  const cols = [
    { label: 'Description', x: PAGE_MARGIN, w: 150, align: 'left' as const },
    { label: 'Qty', x: PAGE_MARGIN + 155, w: 40, align: 'right' as const },
    { label: 'Unit Price', x: PAGE_MARGIN + 200, w: 75, align: 'right' as const },
    { label: 'VAT Rate', x: PAGE_MARGIN + 280, w: 55, align: 'right' as const },
    { label: 'VAT Amt', x: PAGE_MARGIN + 340, w: 70, align: 'right' as const },
    { label: 'Total', x: PAGE_MARGIN + 415, w: 80, align: 'right' as const },
  ];

  // Header row
  const headerY = doc.y;
  doc
    .rect(PAGE_MARGIN, headerY, CONTENT_WIDTH, 18)
    .fill(COLOR_HEADER_BG);

  doc.font(FONT_BOLD).fontSize(8).fillColor(COLOR_PRIMARY);
  for (const col of cols) {
    doc.text(col.label, col.x, headerY + 5, { width: col.w, align: col.align });
  }

  doc.y = headerY + 22;

  // Data rows
  doc.font(FONT_REGULAR).fontSize(8).fillColor(COLOR_SECONDARY);
  for (const item of invoice.items) {
    const rowY = doc.y;

    // Check if we need a new page
    if (rowY > 720) {
      doc.addPage();
      doc.y = PAGE_MARGIN;
    }

    const currentY = doc.y;
    doc.text(item.description, cols[0].x, currentY, { width: cols[0].w, align: cols[0].align });
    doc.text(String(item.quantity), cols[1].x, currentY, { width: cols[1].w, align: cols[1].align });
    doc.text(formatCurrency(item.unitPrice), cols[2].x, currentY, { width: cols[2].w, align: cols[2].align });
    doc.text(`${(item.vatRate * 100).toFixed(1)}%`, cols[3].x, currentY, { width: cols[3].w, align: cols[3].align });
    doc.text(formatCurrency(item.vatAmount), cols[4].x, currentY, { width: cols[4].w, align: cols[4].align });
    doc.text(formatCurrency(item.lineTotal), cols[5].x, currentY, { width: cols[5].w, align: cols[5].align });

    doc.y = currentY + 14;

    // Light separator line
    doc
      .moveTo(PAGE_MARGIN, doc.y)
      .lineTo(PAGE_MARGIN + CONTENT_WIDTH, doc.y)
      .strokeColor(COLOR_LINE)
      .lineWidth(0.5)
      .stroke();

    doc.y += 4;
  }

  // Totals
  doc.moveDown(0.5);
  const totalsX = PAGE_MARGIN + 340;
  const totalsW = 155;

  doc.font(FONT_REGULAR).fontSize(9).fillColor(COLOR_SECONDARY);
  renderTotalRow(doc, 'Subtotal:', formatCurrency(invoice.subtotal), totalsX, totalsW);
  renderTotalRow(doc, 'Total VAT:', formatCurrency(invoice.totalVat), totalsX, totalsW);

  doc.font(FONT_BOLD).fontSize(10).fillColor(COLOR_PRIMARY);
  renderTotalRow(doc, 'TOTAL:', formatCurrency(invoice.total), totalsX, totalsW);

  doc.moveDown(1);
}

function renderTotalRow(
  doc: PDFKit.PDFDocument,
  label: string,
  value: string,
  x: number,
  w: number,
): void {
  const y = doc.y;
  doc.text(label, x, y, { width: 70, align: 'left' });
  doc.text(value, x + 70, y, { width: w - 70, align: 'right' });
  doc.y = y + 14;
}

function renderVatBreakdown(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  if (invoice.vatBreakdown.length === 0) return;

  doc
    .font(FONT_BOLD)
    .fontSize(10)
    .fillColor(COLOR_PRIMARY)
    .text('VAT Breakdown', PAGE_MARGIN, doc.y);

  doc.moveDown(0.3);

  doc.font(FONT_REGULAR).fontSize(8).fillColor(COLOR_SECONDARY);
  for (const entry of invoice.vatBreakdown) {
    const rateLabel = entry.rateType === 'standard'
      ? `${(entry.rate * 100).toFixed(1)}% (Standard)`
      : entry.rateType === 'zero-rated'
        ? '0.0% (Zero-rated)'
        : '0.0% (Exempt)';

    doc.text(
      `${rateLabel}  —  Taxable: ${formatCurrency(entry.taxableAmount)}  |  VAT: ${formatCurrency(entry.vatAmount)}`,
      PAGE_MARGIN,
      doc.y + 1,
    );
  }

  doc.moveDown(1);
}

function renderPaymentInstructions(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  doc
    .font(FONT_BOLD)
    .fontSize(10)
    .fillColor(COLOR_PRIMARY)
    .text('Payment Instructions', PAGE_MARGIN, doc.y);

  doc.moveDown(0.3);
  doc
    .font(FONT_REGULAR)
    .fontSize(8)
    .fillColor(COLOR_SECONDARY)
    .text(`Please remit ${formatCurrency(invoice.total)} (${invoice.currency}) to ${invoice.seller.name}.`);

  if (invoice.dueDate) {
    doc.text(`Payment is due by ${formatDate(invoice.dueDate)}.`);
  }

  if (invoice.notes) {
    doc.moveDown(0.3);
    doc.font(FONT_REGULAR).fontSize(8).fillColor(COLOR_SECONDARY);
    doc.text(`Notes: ${invoice.notes}`);
  }

  doc.moveDown(1);
}

function renderFooter(doc: PDFKit.PDFDocument, invoice: Invoice): void {
  // Separator line
  doc
    .moveTo(PAGE_MARGIN, doc.y)
    .lineTo(PAGE_MARGIN + CONTENT_WIDTH, doc.y)
    .strokeColor(COLOR_LINE)
    .lineWidth(0.5)
    .stroke();

  doc.y += 8;

  doc
    .font(FONT_REGULAR)
    .fontSize(7)
    .fillColor(COLOR_SECONDARY)
    .text(
      'This invoice is issued in compliance with the Nigeria Tax Act (NTA) 2025 and VAT Act as amended. ' +
      'VAT is charged at the applicable rate per NTA 2025 Schedule 1 (zero-rated), Schedule 2 (exempt), ' +
      'or the standard rate of 7.5%. All amounts are in Nigerian Naira (NGN) unless otherwise stated.',
      PAGE_MARGIN,
      doc.y,
      { width: CONTENT_WIDTH },
    );

  doc.moveDown(0.5);
  doc.text(
    `Generated by ngtaxkit — Invoice ${invoice.invoiceNumber} | ${invoice.seller.tin}`,
    PAGE_MARGIN,
    doc.y,
    { width: CONTENT_WIDTH, align: 'center' },
  );
}
